source/validate_datarules.cpp

```diff
@@ -31,42 +31,43 @@ using libspirv::ValidationState_t;
 
 namespace {
 
-// vector types can only have 2 or 3 or 4 components.
-// if Vector16 capability is added, 8 and 16 components are also allowed.
+// Validates that the number of components in the vector type is legal.
+// Vector types can only be parameterized as having 2, 3, or 4 components.
+// If the Vector16 capability is added, 8 and 16 components are also allowed.
 spv_result_t ValidateNumVecComponents(ValidationState_t& _,
                                       const spv_parsed_instruction_t* inst) {
   if (inst->opcode == SpvOpTypeVector) {
-    const std::string opcode_string =
-        spvOpcodeString(static_cast<SpvOp>(inst->opcode));
-
-    // operand 1 specifies the number of components in the vector.
-    uint32_t num_components = inst->words[inst->operands[2].offset];
-    if (num_components == 2 || num_components == 3 || num_components == 4)
+    // operand 2 specifies the number of components in the vector.
+    const uint32_t num_components = inst->words[inst->operands[2].offset];
+    if (num_components == 2 || num_components == 3 || num_components == 4) {
       return SPV_SUCCESS;
-    else if (num_components == 8 || num_components == 16)
-      if (_.HasCapability(SpvCapabilityVector16))
+    } else if (num_components == 8 || num_components == 16) {
```

Comment:
I just noticed that this doesn't have to be an "else" clause, since the 2,3,4 "if" case ends in a return.
So you could have structured it as follows, which is simpler:
```
  if ( 2,3,4...) { return SPV_SUCCESS; }
  if ( 8, 17 ) {
     if ( have vector16 capability ) return success
     special error
  }
  generic error
```